### ISSUE TITLE
`spack external find` for non-executables 

### ISSUE THREAD

**Issue by sethrj** [LLNL/spack#19364]
Add support for `spack external find` for packages that don't have an associated binary executables, using environment variables as hints where appropriate or perhaps [CMake package files](https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure).

### Rationale

There are data-only packages such as `geant4-data` that are useful to have preinstalled as a collection; and other packages (such as a few internal ones in use at ORNL) that can't be downloaded and thus *have* to be installed as externally found packages. The current mechanism for `spack external find` only looks for executables. Most nuclear (and HEP) codes that use external data files have an option for looking up data paths with an environment variable (although some assume that the only data you would ever need is for their application so they choose names like `$DATA` and `$DATAPATH`, but some [geant4] are better and scope their variables like `$G4ENDL`). Currently I have to add these to the `packages.yaml` by hand.

### Description

Packages that don't have executables should support `spack external find` via another mechanism, whether it's based on environment variables or some other more generic approach besides executable paths.

(I'd be happy to help contribute to this feature, but I want to coordinate as discussed in slack.)

**Comment by alalazo** [comment]
@robertu94

**Comment by robertu94** [comment]
One other possible source of the information could be from header files using libffi and/or using the C foreign function interface to call into a library. I have some code that does this for a library that I wrote if it would be useful. 

A few other use cases that I don’t currently have would be would be getting packages from Julia or Python. This may work better for Julia since packages are referred to via a hash

**Comment by wdconinc** [comment]
The current external detection relies on executables, likely to some extent because that is a limited set of paths that can be retrieved easily on all supported systems (`lib/spack/spack/detection/path.py`, `executables_in_path`). Including all files on the system, even those outside of search paths, is likely unpractical. Therefore:
- extending this to environment variables seems equally (even more) limited than executables, so not a priori excluded,
- extending this to libraries may be limited enough to get as well (essentially `ldconfig -p`, ~2500 entries on my system, fewer than executables) (note this would miss all rpath linked libraries since recursing through rpaths of libraries is likely excluded for performance reasons),
- extending this to headers may be too much, in part because there is no equivalent to ldconfig or LD_LIBRARY_PATH for includes outside of system.

So, I'm thinking an extension of `lib/spack/spack/detection/path.py` to `lib/spack/spack/detection/env.py` may be the lowest hanging fruit here, based not on at list of simple regexes for executables but a dict of variable name/value regex. A package would list therefore, e.g. (admittedly taking from my own implementation in the geant4-data PR):
```python
    environment_variables = { 'G4NDL': r'^(?P<prefix>.*?)/share/data/G4NDL(?P<version>.*?)$' }
```
and implement `determine_spec_details(prefix, exes_in_prefix)` as before. The requirement on the regex could be that prefix must be the first captured match, and remaining named captures would be passed as kwargs to `determine_spec_details`. `exes_in_prefix` would be None (or, I guess in a parallel world it would be turned into `vars_in_environment`, but we can't overload `determine_spec_details`). `determine_spec_details` remains responsible for figuring out variants from the specified `prefix`.

In parallel construction similar to `path.py` this would include:
- `def variables_in_environment()`: `"""Get the dictionary of all variables available in the current environment."""` (i.e. strip out functions and multilines, deal with windows etc)
- `def by_environment_variable(packages_to_check)`: `"""Return the list of packages that have been detected, searching by environment."""` (for each package with a found environment variable pre-compile the regex and extract prefix and version)

Thoughts?

**Comment by trws** [comment]
So, I've been thinking about this a lot.  Adding the actual feature is nearly trivial for huge classes of packages, especially those that support pkg-config and/or cmake Find* files, but there's a significant side question: How do we deal with packages that have both, and installations of only one or the other

It's a problem we already have, that causes builds to fail and all kinds of issues if, for example, someone does a find and gets a python with no `-dev` package installed.  There needs to be some way to mark the difference, or have the find fail for partial packages.

**Comment by alalazo** [comment]
Adding a reference to #28005 by @bvanessen since it is closely related to this topic

**Comment by robertu94** [comment]
@trws IMHO, this doesn't need to be complicated.  We can simply extend the current functionality of `executables` to be a list of files that can be found.  Maybe something like:

```python
external_markers = [PkgConfig("lzma"), CMakeTarget("LibLZMA", config=True), AllOf([Library(r"^liblzma.so.(\d\(.\d\.\d.)?)$"), Executable(r"^xz$"), Header("^lzma.h$")], my_determine_version)]
```

or

```python
external_markers = [AllOf([KernelModule("nvidia"), KernelModule("nvidia_uvm"), KernelModule("nvidia_drm"),] lambda x: run(f"modinfo {x[0]}").output)]
```

These would be searched in order.

In these examples, Executable retains the current usage of `executables` and uses the existing meanings of `determine_version`.

PkgConfig("lzma") uses the normal pkgconfig search path as if `pkgconfig` was used.

`CMakeTarget("LibLZMA", config=True)` looks for a CMake Config style package (as opposed to a FindPackage style package) using the normal cmake package search mechanisms.

The `AllOf(search_method_list, files_to_spec_fn)` handles the case of a devel package being missing if a pkg-config file doesn't provide it.  IMHO, if a distribution provides a `FooConfig.cmake` or `foo.pc` but not the corresponding library that tends to suggest a broken package, but using something like `AllOf()` can work arround that when it's needed.

The latter two would need some logic to ensure that installed cmake could find the external package when it goes to search.  Fore example, user calls `spack external find xz`, and locates it via CMake, However the user then needs a newer version of cmake than the version was used to external locate the package.  In my mind this problem can be handled in 1 of 2 ways: 1st by adding the appopriate paths to `CMAKE_PREFIX_PATH` corresponding to where the external package was found. 2nd simply depend on the external cmake and prevent a built cmake from using it (which depending on the package, may be a worse experience).

The second example solves a problem that I have had recently, which is that `cuda` depends on a certain version of the `nvidia.ko` kernel driver, and the newest `cuda` package by spack may not be compatible with the driver package that is installed on the machine.  This is a big pain for me because users aren't always aware of these differences, but may be the cuda maintainers know a better way to address this @Rombur  and @ax3l ?

As for the actual class to implement this I could envision something as simple as

```python
@dataclasses.dataclass(frozen=True)
class Candidate:
      prefix: Path

class LibraryCandidate(Candidate):
      path: Path

class PkgConfigCandiate(Candidate):
     @property
     def version(self):
           pass

     @property
     def cflags(self)
           pass

# etc.

class ExternalSearchMethod(meta=abc.ABCMeta):
     @abc.abstractmethod
     def identify_candiates(self) -> List[Canidate]:
           pass

     @abc.abstractmethod
     def determine_spec_details(self, candidate: Canidate) -> Spec:
           pass
```

**Comment by alalazo** [comment]
Another PR that seems related to this issue https://github.com/spack/spack/pull/29205

**Comment by trws** [comment]
That could be a step in the right direction.

@robertu94, I agree a design like that should work well, and that in general that a pkg-config file without a library is a broken package, but it's easy to have:

1) just the binary/executable, due to say a static install
2) just the library, no headers or executable, if it was pulled in as a lib dependency through say apt
3) all but the executable, due to a lib<whatever>-dev package but lacking the actual executable package

The question we haven't answered is what to do in these cases.  Right now, I'm tempted to say that if *any* of the checks fails, the package should be considered not found, otherwise we have a package that will be broken for some kind of dependency.  I really don't like that solution, because it makes it much harder to use in some cases, but I haven't come up with a better idea short of having a way to express in the database what components exist, or what purposes a given dependency is appropriate for.

**Comment by robertu94** [comment]
@trws I completely agree that these need to be addressed.  I think that a proof of concept that requires all of them is the shortest path forward, and one that doesn't bind our hands later.  A more robust implementation might expand our concept of an external package to be a link-only, run-only, and build-only external package.  In this case

1. the binary provides a run-time only external dependency maybe with a `+executable` variant
2. the library provides a run-time and link-time dependency maybe with a `+lib` variant
3. the last provides a run-time with a `~executable` variant

The default package config implementation might assume `build+link` and `+lib` with the ability to override this if it makes sense.

You also could make a list like `[AllOf(PkgConfig(), Executable), PkgConfig()]` where the first one turns on the `+executable` variant.

**Comment by sethrj** [comment]
For the geant4 data case (and other use cases I have in mind, also transport data) there are no exe files at all, not libraries. Perhaps we could start from there?

**Comment by trws** [comment]
I suppose that has less scope for failure, so sure? The trick we have is making sure the detection stuff can cause the existing exec detection to fail when it should.

The smallest change to do something with this would be to do something like list `ls` as the executable for geant4 and then see if the files exist.  We really don't even need any extra functionality to prototype it, but if you could choose what would it look like?  How do you detect the version for a data-only package like that for example?

**Comment by robertu94** [comment]
@sethrj For the use case you have, we might create some thing like the following:

```python
import os

class G4endlData(Package)
      @staticmethod
      def determine_version_of_data() -> Spec:
            pass

      external_markers = [DataFile("list_of_data2.dat"], prefix_hints=[os.environ.get("G4ENDL", "/"], G4endlData.determine_version_of_data))]
```

I think we should resist the urge to only fix this for only for data files and think about how to fix this for cmake and pkg-config -- even if we don't implement the solutions together.  I do think I agree with @trws that when implemented, we should make the current external packages that can be broken because of missing `-dev` packages fail without them.  That seems to lead us to implementing what I have described as:

+ Header
+ Library
+ Executable
+ DataFile
+ AllOf

In the first pass.  So far I think we could leave 

+ PkgConfig
+ KernelModule
+ CMakeTarget

for later even though I think they are extremely useful.  These have some complexities (what to do when PKG_SEARCH_PATH is changed, how to handle a newer pkg-config when the external found the package with an older one, etc...) that may be nuanced to get right.  However @alalazo linked a PR #29205 that implements PkgConfig, so maybe it's simpler than I estimate.  Does it's author @e4t or its assignee @scheibelp have opinions on this.

I'm also curious what @adamjstewart thinks of this.  I could imagine a `PythonPip` external search method may also be possible if we design this correctly.

**Comment by adamjstewart** [comment]
I wouldn't recommend using external packages for Python libraries, it's currently pretty broken. Because external packages don't register their dependencies in the spec, external Python libs don't actually depend on or extend Python, and so they don't get added to `PYTHONPATH` and may have other problems as well. See #9149.

**Comment by trws** [comment]
I actually might go for pkg-config before headers and libraries as independent things, spack and other packaging ecosystems already provide these, and if they are there, the headers and lib are there or the install is broken and there's a known version available as part of it along with a way to query everything we need to be able to actually use it.  The others are conceptually easier individually, but far harder to compose into something usable.  That said, a top-level thing like `external_markers` that lets us extend this generally and handle multiple things sounds really good.  I didn't think about how tied to executables our `determine_*` methods were until I was reading the PR for pkg-config, but they really really are.  Maybe shoot for smalest possible wrapper that will make that work, with just a File implementation or some such thing so we can re-work pkg-config on top of it and start building the other stuff?

BTW: @asarkar-parsys, this may be relevant
